src/network/networkd-dhcp-static-lease.c

```diff
@@ -0,0 +1,269 @@
+
+        return 0;
+}
+
+int dhcp_static_lease_unset_client_id(DHCPStaticLease *l) {
+        assert_return(l, -EINVAL);
+
+        l->client_id.data = mfree(l->client_id.data);
+        l->client_id.length = 0;
+
+        return 0;
+}
+
+int dhcp_static_lease_section_verify(DHCPStaticLease *l) {
+        assert_return(l, -EINVAL);
+
+        if (section_is_invalid(l->section))
+                return -EINVAL;
+
+        if(!(l->client_id.data && l->client_id.length)) {
+                assert(l->section);
+
+                return log_warning_errno(SYNTHETIC_ERRNO(EINVAL),
+                                         "%s: DHCPStaticLease section without MACAddress= field configured. "
+                                         "Ignoring [DHCPStaticLease] section from line %u.",
+                                         l->section->filename, l->section->line);
+        }
+
+        return 0;
+}
+
+static int lease_new_static(Network *network, const char *filename, unsigned section_line, DHCPStaticLease **ret) {
+        _cleanup_(network_config_section_freep) NetworkConfigSection *n = NULL;
+        _cleanup_(dhcp_static_lease_freep) DHCPStaticLease *static_lease = NULL;
+        int r;
+
+        assert(network);
+        assert(ret);
+        assert(!!filename == (section_line > 0));
+
+        if (filename) {
+                r = network_config_section_new(filename, section_line, &n);
+                if (r < 0)
+                        return r;
+
+                static_lease = hashmap_get(network->dhcp_static_leases_by_section, n);
+                if (static_lease) {
+                        *ret = TAKE_PTR(static_lease);
+                        return 0;
+                }
+        }
+
+        r = dhcp_static_lease_new(&static_lease);
+        if (r < 0)
+                return r;
+
+        static_lease->network = network;
+        LIST_APPEND(dhcp_static_leases, network->dhcp_static_leases, static_lease);
+        network->n_dhcp_static_leases++;
+
+        if (filename) {
+                static_lease->section = TAKE_PTR(n);
+                r = hashmap_ensure_allocated(&network->dhcp_static_leases_by_section, &network_config_hash_ops);
+                if (r < 0)
+                        return r;
+
+                r = hashmap_put(network->dhcp_static_leases_by_section, static_lease->section, static_lease);
+                if (r < 0)
+                        return r;
+        }
+
+        *ret = TAKE_PTR(static_lease);
+
+        return 0; 
+}
+
+int config_parse_dhcp_static_lease_address(
+                const char *unit,
+                const char *filename,
+                unsigned line,
+                const char *section,
+                unsigned section_line,
+                const char *lvalue,
+                int ltype,
+                const char *rvalue, 
+                void *data,
+                void *userdata) {
+
+        Network *network;
+        DHCPStaticLease *current_lease = NULL;
```

Comment:
Please introduce `dhcp_static_lease_free_or_set_invalidp()` or something and use it with `_cleanup_` attribute. See e.g. networkd-address.c.